joomla/joomla-cms
# Add 'subfields' custom field type

### Summary of Changes

This pull request adds the possibility to create custom fields of a new type `subfields`.

Before this PR, it was only possible to add "basic" types as custom fields to articles, contacts etc, whereas "basic" types would be "text", "media", "integer", etc.

This PR adds the possibility to add a `subfields` custom field to an article, contact etc., whereby the idea behind `subfields` is that a user can create a list of sub fields that his custom field should contain, and those sub fields will then be presented to the user in the backend as a possibly repeatable list.

### Testing Instructions

Install the patch, go in the backend to e.g. `Content` -> `Fields` -> `New`, select `Type=Subfields`, set `Title=test123`, under `Fields` add a simple `Text` field with name and label `test1`, and  another simple `Text` field with name and label `test2`. Save & close the custom field.

Go to `Content` -> `Articles` -> `New`, switch to tab `Fields`, and check whether the new `subfields` custom field is working as it should (so displaying a repeatable version of the configured fields). Insert an Article Title and something in the custom field, and save & close the article. Afterwards go to the frontend and check whether the article and its content is being displayed correctly.

### Expected result

Custom field type `subfields` can be edited correctly, its content can be added correctly in the backend when editing/creating an article/contact/etc, and the custom field values will be correctly displayed in the frontend.

### Documentation Changes Required

I have written a small documentation which mainly targets developers and describes the needed actions to make a layout override for the rendering of the `subfields` value. This documentation can be enhanced by everyone interested to describe the main functionality. https://docs.joomla.org/Custom_fields_type:_Subfields

### Additional information

This PR is based on #17552 

I created this PR against the 3.10-dev branch, also because I would like to have enough time to work on possible feedback coming in (not against staging, but directly against 3.10). But can we all please work together so this really ends up in 3.10, and also in 4.0? I am very willing to put much work into this, but I am dependent on you guys' feedback. So please, tell me! :slightly_smiling_face:
